pouta_blueprints/drivers/provisioning/docker_driver.py

```diff
@@ -0,0 +1,542 @@
+        self.logger.debug("do_deprovision %s" % instance_id)
+
+        ap = self._get_ap()
+
+        pbclient = ap.get_pb_client(token, self.config['INTERNAL_API_BASE_URL'], ssl_verify=False)
+        instance = pbclient.get_instance_description(instance_id)
+
+        try:
+            proxy_route = instance['instance_data']['proxy_route']
+            ap.proxy_remove_route(proxy_route)
+        except KeyError:
+            self.logger.info("do_deprovision: No proxy route in instance data for %s" % instance_id)
+
+        if instance['state'] == 'deleted':
+            self.logger.debug("do_deprovision: instance already deleted %s" % instance_id)
+            return
+
+        try:
+            docker_url = instance['instance_data']['docker_url']
+        except KeyError:
+            self.logger.info('no docker url for instance %s, assuming provisioning has failed' % instance_id)
+            return
+
+        dc = ap.get_docker_client(docker_url)
+
+        container_name = instance['name']
+
+        try:
+            dc.remove_container(container_name, force=True)
+        except APIError as e:
+            if e.response.status_code == 404:
+                self.logger.info('no container found for instance %s, assuming already deleted' % instance_id)
+            else:
+                raise e
+
+        self.logger.debug("do_deprovision done for %s" % instance_id)
+
+    @locked('%s/docker_driver_housekeep' % DD_RUNTIME_PATH)
+    def do_housekeep(self, token):
+        return self._do_housekeep(token, int(time.time()))
+
+    # noinspection PyUnusedLocal
+    def _do_housekeep(self, token, cur_ts):
+        ap = self._get_ap()
+
+        # in shutdown mode we remove the hosts as soon as no instances are running on them
+        shutdown_mode = self.config['DD_SHUTDOWN_MODE']
+        if shutdown_mode:
+            self.logger.info('do_housekeep(): in shutdown mode')
+
+        # fetch host data
+        hosts = self._get_hosts(cur_ts)
+
+        self._update_host_lifetimes(hosts, shutdown_mode, cur_ts)
+
+        # count allocated slots
+        num_allocated_slots = sum(x['num_reserved_slots'] for x in hosts)
+
+        # find active hosts
+        active_hosts = [x for x in hosts if x['state'] == DD_STATE_ACTIVE]
+
+        # find current free slots
+        num_free_slots = sum(x['num_slots'] - x['num_reserved_slots'] for x in active_hosts)
+
+        # find projected free slots (only take active hosts with more than a minute to go)
+        num_projected_free_slots = sum(
+            x['num_slots'] - x['num_reserved_slots'] for x in active_hosts
+            if x['lifetime_left'] > DD_HOST_LIFETIME_LOW
+        )
+
+        self.logger.info(
+            'do_housekeep(): active hosts: %d, free slots: %d now, %d projected for near future' %
+            (len(active_hosts), num_free_slots, num_projected_free_slots)
+        )
+
+        # priority one: find just spawned hosts, prepare and activate those
```

Comment:
Probably define each "priority handler" in its own method returning value that indicates whether the housekeeping should proceed or return. Now the function can return from multiple places and multiple indentation levels, making it quite hard to read in a single read.